README.markdown
# ZK #

![Build Status](https://github.com/zk-ruby/zk/actions/workflows/build.yml/badge.svg)

ZK is an application programmer's interface to the Apache [ZooKeeper][] server. It is based on the [zookeeper gem][] which is a multi-Ruby low-level driver. Currently MRI 1.8.7, 1.9.2, 1.9.3, REE, and JRuby are supported. Rubinius 2.0.testing is supported-ish (it's expected to work, but upstream is unstable, so YMMV).

ZK is licensed under the [MIT][] license.

The key place to start in the documentation is with ZK::Client::Base ([rubydoc.info][ZK::Client::Base], [local](/docs/ZK/Client/Base)).

See the [RELEASES][] file for information on what changed between versions.

This library is heavily used in a production deployment and is actively developed and maintained.

Development is sponsored by [Snapfish][] and has been generously released to the Open Source community by HPDC, L.P.

[ZK::Client::Base]: http://rubydoc.info/gems/zk/ZK/Client/Base
[ZooKeeper]: http://zookeeper.apache.org/ "Apache ZooKeeper"
[zookeeper gem]: https://github.com/zk-ruby/zookeeper "zookeeper gem"
[MIT]: http://www.gnu.org/licenses/license-list.html#Expat "MIT (Expat) License"
[Snapfish]: http://www.snapfish.com/ "Snapfish"
[RELEASES]: https://github.com/zk-ruby/zk/blob/master/RELEASES.markdown

## What is ZooKeeper? ##

ZooKeeper is a multi-purpose tool that is designed to allow you to write code that coordinates many nodes in a cluster. It can be used as a directory service, a configuration database, and can provide cross-cluster [locking][], [leader election][], and [group membership][] (to name a few). It presents to the user what looks like a distributed file system, with a few important differences: every node can have children _and_ data, and there is a 1MB limit on data size for any given node. ZooKeeper provides atomic semantics and a simple API for manipulating data in the heirarchy.

One of the most useful aspects of ZooKeeper is the ability to set "[watches][]" on nodes. This allows one to be notified when a node has been deleted, created, changed, or has had its list of child znodes modified. The asynchronous nature of these watches enables you to write code that can _react_ to changes in your environment without polling and busy-waiting.

Znodes can be _ephemeral_, which means that when the connection that created them goes away, they're automatically cleaned up, and all the clients that were watching them are notified of the deletion. This is an incredibly useful mechanism for providing _presence_ in a cluster ("which of my thingamabobers are up?). If you've ever run across a stale pid file or lock, you can imagine how useful this feature can be.

Znodes can also be created as _sequence_ nodes, which means that beneath a given path, a node can be created with a given prefix and assigned a unique integer. This, along with the _ephemeral_ property, provide the basis for most of the coordination classes such as [groups][] and [locks][].

ZooKeeper is easy to deploy in a [Highly Available][ha-config] configuration, and the clients natively understand the clustering and how to resume a session transparently when one of the cluster nodes goes away.

[watches]: http://zookeeper.apache.org/doc/current/zookeeperProgrammers.html#ch_zkWatches
[locking]: http://zookeeper.apache.org/doc/current/recipes.html#sc_recipes_Locks
[leader election]: http://zookeeper.apache.org/doc/current/recipes.html#sc_leaderElection
[group membership]: http://zookeeper.apache.org/doc/current/recipes.html#sc_outOfTheBox
[ha-config]: http://zookeeper.apache.org/doc/current/zookeeperAdmin.html#sc_CrossMachineRequirements "HA config"
[groups]: https://github.com/zk-ruby/zk-group
[locks]: http://rubydoc.info/gems/zk/ZK/Locker


## What does ZK do that the zookeeper gem doesn't?

The [zookeeper gem][] provides a low-level, cross platform library for interfacing with ZooKeeper. While it is full featured, it only handles the basic operations that the driver provides. ZK implements the majority of the [recipes][] in the ZooKeeper documentation, plus a number of other conveniences for a production environment. ZK aims to be to Zookeeper, as Sequel or ActiveRecord is to the MySQL or Postgres drivers (not that ZK is attempting to provide an object persistence system, but rather a higher level API that users can develop applications with).

ZK provides:

* 	a robust lock implementation (both shared and exclusive locks)
* 	a leader election implementation with both "leader" and "observer" roles
* 	a higher-level interface to the ZooKeeper callback/watcher mechanism than the [zookeeper gem][] provides
* 	a simple threadpool implementation
* 	a bounded, dynamically-growable (threadsafe) client pool implementation
* 	a recursive Find class (like the Find module in ruby-core)
* 	unix-like rm\_rf and mkdir\_p methods
* 	an extension for the [Mongoid][] ORM to provide advisory locks on mongodb records

In addition to all of that, I would like to think that the public API the ZK::Client provides is more convenient to use for the common (synchronous) case. For use with [EventMachine][] there is [zk-eventmachine][] which provides a convenient API for writing evented code that uses the ZooKeeper server.

[recipes]: http://zookeeper.apache.org/doc/current/recipes.html
[Mongoid]: http://mongoid.org/
[EventMachine]: https://github.com/eventmachine/eventmachine
[zk-eventmachine]: https://github.com/zk-ruby/zk-eventmachine

## Release info / Changelog

See the [RELEASES][] page for more info on features and bugfixes in each release.

## Caveats

ZK strives to be a complete, correct, and convenient way of interacting with ZooKeeper. There are a few things to be aware of:

* In versions < 0.9 there is only *one* event dispatch thread. It is *very important* that you don't block the event delivery thread. In 1.0, there is one delivery thread by default, but you can adjust the level of concurrency, allowing more control and convenience for building your event-driven app.

* ZK uses threads. You will have to use synchronization primitives if you want to avoid getting hurt. There are use cases that do not require you to think about this, but as soon as you want to register for events, you're using multiple threads.

* If you're not familiar with developing solutions with zookeeper, you should read about [sessions][] and [watches][] in the Programmer's Guide. Even if you *are* familiar, you should probably go read it again.

* It is very important that you not ignore connection state events if you're using watches.

* _ACLS: HOW DO THEY WORK?!_  ACL support is mainly faith-based now. I have not had a need for ACLs, and the authors of the upstream [twitter/zookeeper][] code also don't seem to have much experience with them/use for them (purely my opinion, no offense intended). If you are using ACLs and you find bugs or have suggestions, I would much appreciate feedback or examples of how they *should* work so that support and tests can be added.

* ZK::Client supports asynchronous calls of all basic methods (get, set, delete, etc.) however these versions are kind of inconvenient to use. For a fully evented stack, try [zk-eventmachine][], which is designed to be compatible and convenient to use in event-driven code.

[twitter/zookeeper]: https://github.com/twitter/zookeeper
[async-branch]: https://github.com/zk-ruby/zk/tree/dev%2Fasync-conveniences
[chroot]: http://zookeeper.apache.org/doc/current/zookeeperProgrammers.html#ch_zkSessions
[YARD]: http://yardoc.org/
[sessions]: http://zookeeper.apache.org/doc/current/zookeeperProgrammers.html#ch_zkSessions
[watches]: http://zookeeper.apache.org/doc/r3.3.5/zookeeperProgrammers.html#ch_zkWatches

## Users

* [papertrail](http://papertrailapp.com/): Hosted log management service
* [redis\_failover](https://github.com/ryanlecompte/redis_failover): Redis client/server failover managment system
* [DCell](https://github.com/celluloid/dcell): Distributed ruby objects, built on top of the super cool [Celluloid](https://github.com/celluloid/celluloid) framework.


## Dependencies

* The [zookeeper gem][szk-gem] ([repo][szk-repo]).

* For JRuby, the [slyphon-zookeeper\_jar gem][szk-jar-gem] ([repo][szk-jar-repo]), which just wraps the upstream zookeeper driver jar in a gem for easy installation

[szk-gem]: https://rubygems.org/gems/zookeeper
[szk-repo]: https://github.com/zk-ruby/zookeeper
[szk-repo-bundler]: https://github.com/zk-ruby/zookeeper/tree/dev/gemfile/
[szk-jar-gem]: https://rubygems.org/gems/slyphon-zookeeper_jar
[szk-jar-repo]: https://github.com/zk-ruby/zookeeper_jar

## Contacting the author

* if you really want to, you can also reach me via twitter [@slyphon][]

[@slyphon]: https://twitter.com/#!/slyphon




RELEASES.markdown
This file notes feature differences and bugfixes contained between releases.

### v1.10.0 ###

* Updates dependency on zookeeper gem to 1.5.0 #90 (h/t: @malmckay)
* Move build to Github Actions, drop support for older Rubies (#104) (h/t: @malmckay)
* bump yard revision to ~> 0.9.11 (Fixes #101) (#102) (h/t: @slyphon)
* Add other Ruby versions with currently breaking tests (h/t: @berlincount)
* RSpec 3 support and cleanups (h/t: @nerdrew)
* Interrupt blocked locker (h/t: @lmarburger)

### v1.9.6 ###

* Fixes from @rickypai for ruby 2.2 (#89)
* Remove dependency on logging gem #90 (h/t: @eric)

### v1.9.5 ###

* Really clear hooks when clear! is called #83 (h/t: Liam Stewart)
* implement `add_auth` method to send credentials to zookeeper #86 (h/t: ajf8)

### v1.9.4 ###

* Forward options to underlying connection #69 (h/t: avalanche123)
* Don't check connection state in Locker#assert! - leads to better retry behavior
* allow specifying session-id
* upgrade logging gem dependency


### v1.9.3 ###

* Fix deadlocks between watchers and reconnecting


### v1.9.2 ###

* Fix re-watching znodes after a lost session #72 (reported by kalantar)


### v1.9.1 ###

* Fix re-rewatching children watchers after the parent znode was deleted #68
* Deal with reopening a closed connection properly #70


### v1.9.0 ###

* Semaphores!
* shared/exclusive lock fixes

### v1.8.0 ###

* Added non-exploderating Locker#assert method (issue #48, h/t: johnbellone)

### v1.7.5 ###

* fix for jruby 1.7 (issue #53)

### v1.7.4 ###

* Narsty bug in Locker (#54)

If a locker is waiting on the lock, and a connection interruption occurs (that doesn't render the session invalid), the waiter will attempt to clean up while the connection is invalid, and not succeed in cleaning up its ephemeral. This patch will recognize that the `@lock_path` was already acquired, and just wait on the current owner (ie. it won't create an erroneous *third* lock node). The reproduction code has been added under `spec/informal/two-locks-enter-three-locks-leave.rb`


### v1.7.3 ###

* bug fix for "Callbacks Hash in EventHandlerSubscription::Base gets longer randomly" (#52)

I'd like to point out that the callbacks hash gets longer *deterministically*, depending on what callbacks get registered. This patch will do further cleanup so as not to leave empty arrays littering the EventHandler.

### v1.7.2 ###

* bug fix for "Ephemeral node for exclusive lock not cleaned up when failure happens during lock acquisition" (#51)

### v1.7.1 ###

* Fixes nasty bug "LockWaitTimeout causes lock to be forever unusable" (#49)

The code path in the case of a LockWaitTimeout would skip the lock node cleanup, so a given lock name would become unusable until the timed-out-locker's session went away. This fixes that case and adds specs.

### v1.7.0 ###

* Added Locker timeout feature for blocking calls. (issue #40)

Previously, when dealing with locks, there were only two options: blocking or non-blocking. In order to come up with a time-limited lock, you had to poll every so often until you acquired the lock. This is, needless to say, both inefficient and doesn't allow for fair acquisition.

A timeout option has been added so that when blocking waiting for a lock, you can specify a deadline by which the lock should have been acquired.

```ruby
zk = ZK.new

locker = zk.locker('lock name')

begin
  locker.lock(:wait => 5.0)   # wait up to 5.0 seconds to acquire the lock
rescue ZK::Exceptions::LockWaitTimeoutError
  $stderr.puts "could not acquire the lock in time"
end
```

Also available when using the convenience `#with_lock` methods

```ruby

zk = ZK.new

begin
  zk.with_lock('lock name', :wait => 5.0) do |lock|
    # do stuff while holding lock
  end
rescue ZK::Exceptions::LockWaitTimeoutError
  $stderr.puts "could not acquire the lock in time"
end

```


### v1.6.4 ###

* Remove unnecessary dependency on backports gem
* Fix for use in resque! A small bug was preventing resque from activating the fork hook.

### v1.6.3 ###

* Retry when lock creation fails due to a NoNode exception

### v1.6.2 ###

* Change state call to reduce the chances of deadlocks

One of the problems I've been seeing is that during some kind of shutdown event, some method will call `closed?` or `connected?` which will acquire a mutex and make a call on the underlying connection at the *exact* moment necessary to cause a deadlock. In order to help prevent this, and building on some changes from 1.5.3, we now treat our cached `@last_cnx_state` as the current state of the connection and don't touch the underlying connection object (except in the case of the java driver, which is safe).

### v1.6.1 ###

* Small fixes for zk-eventmachine compatibilty

### v1.6.0 ###

* Locker cleanup code!

When a session is lost, it's likely that the locker's node name was left behind. so for `zk.locker('foo')` if the session is interrupted, it's very likely that the `/_zklocking/foo` znode has been left behind. A method has been added to allow you to safely clean up these stale znodes:

```ruby
ZK.open('localhost:2181') do |zk|
  ZK::Locker.cleanup(zk)
end
```

Will go through your locker nodes one by one and try to lock and unlock them. If it succeeds, the lock is naturally cleaned up (as part of the normal teardown code), if it doesn't acquire the lock, then no harm, it knows that lock is still in use.

* Added `create('/path', 'data', :or => :set)` which will create a node (and all parent paths) with the given data or set its contents if it already exists. It's intended as a convenience when you just want a node to exist with a particular value.

### v1.5.3 ###


lib/zk.rb
require 'rubygems'

require 'zookeeper'

# XXX: after 1.0 we'll need this
# require 'zookeeper/compatibility'

require 'forwardable'
require 'thread'
require 'monitor'
require 'set'
require 'time'
require 'date'

module ZK
  # just like stdlib Monitor but provides the SAME API AS MUTEX, FFS!
  # @private
  class Monitor < Zookeeper::Monitor
  end
end

require 'zk/core_ext'
require 'zk/exceptions'
require 'zk/extensions'
require 'zk/logger'
require 'zk/event'
require 'zk/stat'
require 'zk/subscription'
require 'zk/fork_hook'
require 'zk/threadpool'
require 'zk/threaded_callback'
require 'zk/event_handler_subscription'
require 'zk/event_handler'
require 'zk/node_deletion_watcher'
require 'zk/message_queue'
require 'zk/locker'
require 'zk/election'
require 'zk/mongoid'
require 'zk/client'
require 'zk/pool'
require 'zk/find'

module ZK
  silence_warnings do
    # @private
    ZK_ROOT = File.expand_path('../..', __FILE__).freeze
  end

  unless defined?(KILL_TOKEN) 
    # @private
    KILL_TOKEN = Object.new 
  end

  @@logger = nil unless defined? @@logger
 
  @default_host   = 'localhost' unless @default_host
  @default_port   = 2181        unless @default_port
  @default_chroot = ''          unless @default_chroot

  class << self
    # what host should ZK.new connect to when given no options
    attr_accessor :default_host

    # what port should ZK.new connect to when given no options
    attr_accessor :default_port

    # what chroot path should ZK.new connect to when given no options
    attr_accessor :default_chroot
  end

  # The logger used by the ZK library. uses a Logger stderr with Logger::ERROR
  # level. The only thing that should ever be logged are exceptions that are
  # swallowed by background threads.
  #
  # You can change this logger by setting ZK#logger= to an object that
  # implements the stdllb Logger API.
  #
  def self.logger
    @@logger
  end

  # set the ZK logger instance
  def self.logger=(log)
    @@logger = log
  end

  # @private
  def self.default_connection_string
    "#{default_host}:#{default_port}".tap do |str|
      # XXX: this is seriously blech
      str.replace(File.join(str, default_chroot)) if default_chroot != ''
    end
  end

  # Create a new ZK::Client instance. If no arguments are given, the default
  # config of `localhost:2181` will be used. Otherwise all args will be passed
  # to ZK::Client#new
  #
  # if a block is given, it will be yielded the client *before* the connection
  # is established, this is useful for registering connected-state handlers.
  #
  # Since 1.0, if you pass a chrooted host string, i.e. `localhost:2181/foo/bar/baz` this
  # method will create two connections. The first will be short lived, and will create the 
  # chroot path, the second will be the chrooted one and returned to the user. This is
  # meant as a convenience to users who want to use chrooted connections.
  #
  # @note As it says in the ZooKeeper [documentation](http://zookeeper.apache.org/doc/r3.4.3/zookeeperProgrammers.html#ch_gotchas), 
  #   if you are running a cluster: "The list of ZooKeeper servers used by the
  #   client must match the list of ZooKeeper servers that each ZooKeeper
  #   server has. Things can work, although not optimally, if the client list
  #   is a subset of the real list of ZooKeeper servers, but not if the client
  #   lists ZooKeeper servers not in the ZooKeeper cluster."
  #
  # @example Connection using defaults
  #
  #   zk = ZK.new   # will connect to 'localhost:2181' 
  #
  # @example Connection to a single server
  #
  #   zk = ZK.new('localhost:2181')
  #
  # @example Connection to a single server with a chroot (automatically created)
  #
  #   zk = ZK.new('localhost:2181/look/around/you')
  #
  # @example Connection to multiple servers (a cluster)
  #   
  #   zk = ZK.new('server1:2181,server2:2181,server3:2181')
  #
  # @example Connection to multiple servers with a chroot (chroot will automatically be creatd)
  #
  #   zk = ZK.new('server1:2181,server2:2181,server3:2181/look/around/you')
  #
  # @example Connection to a single server, assert that chroot path exists, but do not create it
  #
  #   zk = ZK.new('localhost:2181/look/around/you', :chroot => :check)
  #
  # @example Connection to a single server, use a chrooted connection, do not check for validity, do not create
  # 
  #   zk = ZK.new('localhost:2181/look/around/you', :chroot => :do_nothing)
  #
  # @example Connection to a single server, chroot path specified as an option
  #
  #   zk = ZK.new('localhost:2181', :chroot => '/look/around/you')
  #
  # @overload new(connection_str, opts={}, &block)
  #   @param [String] connection_str A zookeeper host connection string, which
  #     is a comma-separated list of zookeeper servers and an optional chroot
  #     path.
  #
  #   @option opts [:create,:check,:do_nothing,String] :chroot (:create) if a chrooted
  #     `connection_str`, `:chroot` can have the following values:
  #
  #     * `:create` (the default), then we will use a secondary (short-lived)
  #     un-chrooted connection to ensure that the path exists before returning
  #     the chrooted connection. 
  #
  #     * `:check`, we will not attempt to create the connection, but rather
  #     will raise a {Exceptions::ChrootPathDoesNotExistError
  #     ChrootPathDoesNotExistError} if the path doesn't exist. 
  #
  #     * `:do_nothing`, we do not create the path and furthermore we do not
  #     perform the check (the `<= 0.9` behavior).
  #
  #     * if a `String` is given, it is used as the chroot path, and we will follow
  #     the same rules as if `:create` was given if `connection_str` also
  #     contains a chroot path, we raise an `ArgumentError`
  #
  #     * if you don't like this for some reason, you can always use
  #     {ZK::Client::Threaded#initialize Threaded.new} directly. You probably
  #     also hate happiness and laughter.
  #
  #   @option opts [:single,:per_callback] :thread (:single) see {ZK::Client::Threaded#initialize} 
  #     for a discussion of what these options mean
  #
  #   @raise [ChrootPathDoesNotExistError] if a chroot path is specified,
  #     `:chroot` is `:check`, and the path does not exist.
  #
  #   @raise [ArgumentError] if both a chrooted `connection_str` is given *and* a
  #     `String` value for the `:chroot` option is given
  #
  #
  def self.new(*args, &block)
    opts = args.extract_options!

    chroot_opt = opts.fetch(:chroot, :create)

    args = [default_connection_string]  if args.empty?     # the ZK.new() case

    if args.first.kind_of?(String)
      if new_cnx_str = do_chroot_setup(args.first, chroot_opt)
        args[0] = new_cnx_str
      end
    else
      raise ArgumentError, "cannot create a connection given args array: #{args}"
    end

    opts.delete(:chroot_opt)

    args << opts

    Client.new(*args, &block)
  end

  # Like new, yields a connection to the given block and closes it when the
  # block returns
  def self.open(*args)
    cnx = new(*args)
    yield cnx
  ensure
    if cnx
      cnx.close! 
      cnx.wait_until_closed(30) # XXX: hardcoded here, do not hang forever
    end
  end

  # creates a new ZK::Pool::Bounded with the default options.
  def self.new_pool(host, opts={})
    ZK::Pool::Bounded.new(host, opts)
  end

  # @private
  def self.join(*paths)
    File.join(*paths)
  end

  # @private
  def self.mri_19?
    ruby_19? and not jruby? or rubinius?
  end

  def self.mri_193?
    (RUBY_VERSION == '1.9.3') and not jruby? or rubinius?
  end

  def self.ruby_19?
    (RUBY_VERSION =~ /\A1\.9\.[2-9]\Z/)
  end

  def self.ruby_187?
    (RUBY_VERSION == '1.8.7')
  end

  # @private
  def self.mri_187?
    ruby_187? and not jruby? or rubinius?
  end

  # @private
  def self.jruby?
    defined?(::JRUBY_VERSION)
  end

  # @private
  def self.rubinius?
    defined?(::Rubinius)
  end

  private
    # @return [String] a possibly modified connection string (with chroot info
    #   added)
    #
    def self.do_chroot_setup(cnx_str, chroot_opt=:create)
      # "it should set up the chroot for us," they says. 
      # "it's confusing if it doesn't do that for us," they says.
      # sheesh, look at this...

      host, chroot_path = Client.split_chroot(cnx_str)

      case chroot_opt
      when :do_nothing
        return
      when String
        if chroot_path
          raise ArgumentError, "You cannot give a connection_str with a chroot path (#{cnx_str}) *and* specify a :chroot => #{chroot_opt} too!"
        else
          # ok, cnx_str didn't have a chroot path on it, but the user
          # specified :chroot => '/path'. we'll use that, then
          chroot_path = chroot_opt.dup
          chroot_opt  = :create

          # oh, and return the correct string later
          cnx_str = "#{host}#{chroot_path}"
        end
      when :create, :check
        # no-op, valid options for later
      else
        raise ArgumentError, ":chroot must be one of :create, :check, :do_nothing, or a String, not: #{chroot_opt.inspect}" 
      end

      return cnx_str unless chroot_path  # if by this point, we don't have a chroot_path, then there isn't one to be had

      # make sure the given path is kosher
      Client.assert_valid_chroot_str!(chroot_path)

      open(host) do |zk|                # do path stuff with the virgin connection
        unless zk.exists?(chroot_path)  # someting must be done
          if chroot_opt == :create      # here, let me...
            zk.mkdir_p(chroot_path)     # ...get that for you
          else                          # careful with that axe
            raise Exceptions::ChrootPathDoesNotExistError.new(host, chroot_path)  # ...eugene
          end                                                                               
        end
      end

      cnx_str   # the possibly-modified connection string (with chroot info)
    end
end
